pkg/externalresource/manager/service.go

```diff
@@ -54,5 +167,199 @@ func (s *Service) CreateResource(
 		return nil, st.Err()
 	}
 
+	// Note that the cache is updated AFTER the metastore is updated.
+	if _, exists := s.cache[request.GetResourceId()]; exists {
+		log.L().Panic("Cache is inconsistent",
+			zap.String("resource-id", request.GetResourceId()))
+	}
+	s.cache[request.GetResourceId()] = resourceRecord
+
+	// TODO: handle the case where resourceRecord.Deleted == true
 	return &pb.CreateResourceResponse{}, nil
 }
+
+func (s *Service) GetPlacementConstraint(
+	ctx context.Context,
+	id resourcemeta.ResourceID,
+) (resourcemeta.ExecutorID, bool, error) {
+	if !s.checkAllLoaded() {
+		return "", false, derror.ErrResourceManagerNotReady.GenWithStackByArgs()
+	}
+
+	logger := log.L().WithFields(zap.String("resource-id", id))
+
+	rType, _, err := resourcemeta.ParseResourcePath(id)
+	if err != nil {
+		return "", false, err
+	}
+
+	if rType != resourcemeta.ResourceTypeLocalFile {
+		logger.Info("Resource does not need a constraint",
+			zap.String("resource-id", id), zap.String("type", string(rType)))
+		return "", false, nil
+	}
+
+	if !s.mu.Lock(ctx) {
+		return "", false, errors.Trace(ctx.Err())
+	}
+	defer s.mu.Unlock()
+
+	record, exists := s.cache[id]
+	if !exists {
+		// Note that although we are not doing cache eviction,
+		// a miss is still a possibility given that we might have
+		// a successful write to metastore but due to network problem
+		// we have treated that write as an error.
+		logger.Info("Resource cache miss")
+		var err error
+
+		startTime := time.Now()
+		record, exists, err = s.accessor.GetResource(ctx, id)
```

Comment:
here we may extract a helper function to reduce duplication with QueryResource

Reply:
Yes, good idea.